KAFKA-17309: Fix flaky testCallFailWithUnsupportedVersionExceptionDoesNotHaveConcurrentModificationException

FrankYang0529 opened the pull request:
When closing `KafkaAdminClient`,  the process may hang, because there is active external request. Prepare the response for the request, so the `KafkaAdminClient` can close smoothly.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

frankvicky commented:
I have fetch this PR and loop in local.
It's very stable now.

JDK Version: 21 (Amazon Corretto)
Command I use: 
```bash
I=0;  while ./gradlew clients:test --tests KafkaAdminClientTest.testCallFailWithUnsupportedVersionExceptionDoesNotHaveConcurrentModificationException --rerun --fail-fast; do  (( I=$I+1 )); echo "Completed run: $I"; sleep 1; done
```

![image](https://github.com/user-attachments/assets/84599f33-914e-443d-93f5-df45afd95eff)

frankvicky reviewed (approved):
LGTM

chia7712 commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -8192,6 +8193,9 @@ public void testListClientMetricsResourcesNotSupported() {
     @Test
     public void testCallFailWithUnsupportedVersionExceptionDoesNotHaveConcurrentModificationException() throws InterruptedException {
         try (AdminClientUnitTestEnv env = new AdminClientUnitTestEnv(mockCluster(1, 0))) {
+            // mock AdminMetadataManager for KafkaAdminClient, so closing the default one
```
It seems you don't use the `env` actually. Maybe you can remove the env from this test case and create `MockClient` by yourself.

FrankYang0529 replied:
Yes, updated it. Thanks for the review.

chia7712 reviewed (commented):
@FrankYang0529 thanks for this quick fix.

chia7712 commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -8191,7 +8192,23 @@ public void testListClientMetricsResourcesNotSupported() {
     public void testCallFailWithUnsupportedVersionExceptionDoesNotHaveConcurrentModificationException() throws InterruptedException {
-        try (AdminClientUnitTestEnv env = new AdminClientUnitTestEnv(mockCluster(1, 0))) {
+        Cluster cluster = mockCluster(1, 0);
+        try (MockClient mockCLient = new MockClient(Time.SYSTEM, new MockClient.MockMetadataUpdater() {
```
`mockCLient` -> `mockClient`

chia7712 commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -8206,16 +8223,25 @@ public void testCallFailWithUnsupportedVersionExceptionDoesNotHaveConcurrentModi
+
+                // after handleUnsupportedVersionException, describe cluster use MetadataRequest
+                ClientRequest request = mockCLient.requests().peek();
+                assertNotNull(request.apiKey());
```
both line 8239 and line 8240 could be replaced by `assertEquals(ApiKeys.METADATA, request.apiKey());`

chia7712 commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -8206,16 +8223,25 @@ public void testCallFailWithUnsupportedVersionExceptionDoesNotHaveConcurrentModi
+
+                // clear active external request
+                mockCLient.respondToRequest(request, prepareMetadataResponse(cluster, Errors.NONE));
+                assertDoesNotThrow(() -> result.clusterId().get());
```
`assertEquals(cluster.clusterResource().clusterId(), assertDoesNotThrow(() -> result.clusterId().get()));`

chia7712 reviewed (commented):
@FrankYang0529 thanks for updates

FrankYang0529 committed:
KAFKA-17309: Fix flaky testCallFailWithUnsupportedVersionExceptionDoesNotHaveConcurrentModificationException

Signed-off-by: PoAn Yang <payang@apache.org>

FrankYang0529 commented:
Hi @chia7712, thanks for the review. I addressed all comments. Failed cases in CI can't reproduce on my laptop.

```
./gradlew cleanTest :streams:test --tests RestoreIntegrationTest.shouldInvokeUserDefinedGlobalStateRestoreListener :trogdor:test --tests CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated :core:test --tests PlaintextAdminIntegrationTest.testDescribeTransactions --tests DelegationTokenRequestsTest.testDelegationTokenRequests --tests ZkMigrationFailoverTest.testDriverSkipsEventsFromOlderEpoch :clients:test --tests MemoryRecordsBuilderTest.testBuffersDereferencedOnClose
```

chia7712 reviewed (approved):
LGTM
